docs/reference/docs/data-replication.asciidoc

```diff
@@ -0,0 +1,136 @@
+. Execute the operation locally - i.e., indexing or deleting the relevant document. This will also validate the content of fields
+   and reject if needed (Example: a keyword value is too long for indexing in Lucene).
+. Forward the operation to each replica in the current in-sync copies set. If there are multiple replicas, this is done concurrently.
+. Once all replicas have successfully performed the operation and responded to the primary, the primary acknowledges the successful
+   completion of the request to the client.
+
+==== Failure handling
+
+Many things can go wrong during indexing - disks can get corrupted, nodes can be disconnected from each other or some
+configuration mistake could cause an operation to fail on a replica despite it being successful on the primary. These
+are infrequent but the primary has to respond to them.
+
+In the case that the primary itself fails, a message will be sent to the master about it. The indexing operation will
+wait (up to 1 minute, by [default](link to doc)) for the master to promote one of the replicas to be a new primary.
+The operation will then be forwarded to the new primary for processing. Note that the master also monitors the health
+of the nodes and may decide to pro-actively demote a primary. This typically happens when the node holding the primary
+ is isolated from the cluster by a networking issue. More on this later.
+
+Once the operation has been  successfully performed on the primary, the primary has to deal with potential failures
+when executing it on the replica shards. This may be caused by an actual failure on the replica or due to a network
+issue preventing the operation from reaching the replica (or preventing the replica from responding). All of these
+share the same end result - a replica which is part of the in-sync replica set misses an operation that is about to
+be acknowledged. In order to avoid violating the invariant, the primary sends a message to  the master requesting
+that the problematic shard be removed from the in-sync replica set. Only once this has successfully been done, does
+the primary acknowledge the operation. Note that the master will also instruct another node to start building a new
+shard copy in order to restore the system to a healthy state.
+
+While forwarding an operation to the replicas, the primary will use the replicas to validate that it is still the
+active primary. If the primary has been isolated due to a network partition (or a long GC) it may continue to process
+ incoming indexing operations before realising that it has been demoted. Operations that come from a stale primary
+ will be rejected by the replicas. The primary, following its protocol, will then reach out to the master and will
+ learn that it has been replaced. The operation is then routed to the new primary.
+
+What happens if there are no replicas? This is a valid scenario that can happen due to index configuration or simply
```

Comment:
it's more subtle than that. The problematic scenario is having *two* primaries at the same time. If there are no replicas to promote, the master will think that there is no primary and the primary is free to process requests (it's alone). That "no replicas" doesn't mean "no unassigned in-sync replicas". For the latter the primary will reach out to the master to remove the replica from the in sync set and thus making sure it will never be promoted as another primary.